File: ACM/tree/p2042. 구간 합 구하기.java
import java.util.ArrayList;
import java.util.Scanner;

public class Main {
	
	static long[] tree;
	
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		int M = sc.nextInt();
		int K = sc.nextInt();
		
		long MIN_NUM = Long.MIN_VALUE;
		long MAX_NUM = Long.MAX_VALUE;
		
		int k = 0;
		while (N > (int) Math.pow(2, k)) {
			k++;
		}
		
		// 2 ** 19 = 524,288
		// 2 ** 20 = 1,048,576
		// 2 ** 21 = 2,097,152
		// int 범위 = -2,147,483,648 ~ 2,147,483,647
		int treeSize = (int) Math.pow(2, k+1);
		tree = new long[treeSize];
		
		int dataStartIdx = treeSize / 2;
		// 입력으로 주어지는 모든 수는 long타입 범위
		for (int i=0; i < N; i++) {
			tree[dataStartIdx + i] = sc.nextLong();
		}
		
		int tempStartIdx = dataStartIdx;
		int tempCnt = treeSize / 4;
		
		while (tempCnt > 0) {
			int parentIdx = tempStartIdx / 2;
			for (int i=0; i < tempCnt; i++) {
				tree[parentIdx + i] = tree[tempStartIdx + 2*i] + tree[tempStartIdx + 2*i + 1];
			}
			tempCnt = tempCnt / 2;
			tempStartIdx = tempStartIdx / 2;
		}
		
		for (int i=0; i < M+K; i++) {
			int a = sc.nextInt();
			int b = sc.nextInt();
			long c = sc.nextLong();
			
			int startIndex = dataStartIdx + b - 1;
			if (a == 1) {
				tree[startIndex] = c;
				int parentIdx = startIndex / 2;
				while (parentIdx >= 1) {
					int leftChildIdx = parentIdx * 2;
					int rightChildIdx = leftChildIdx + 1;
					tree[parentIdx] = tree[leftChildIdx] + tree[rightChildIdx];
					parentIdx = parentIdx / 2;
				}
			} else {
				int endIndex = dataStartIdx + (int) c - 1;
				ArrayList<Long> pocket = new ArrayList<Long>();
				while (endIndex >= startIndex) {
					if (startIndex % 2 == 1) {
						pocket.add(tree[startIndex]);
					}
					if (endIndex % 2 == 0) {
						pocket.add(tree[endIndex]);
					}
					startIndex = (startIndex + 1) / 2;
					endIndex = (endIndex - 1) / 2;
				}
				
				long prefixSum = 0;
				for (long element : pocket) {
					prefixSum += element;
				}
				if (prefixSum > MAX_NUM) prefixSum = MAX_NUM;
				else if (prefixSum < MIN_NUM) prefixSum = MIN_NUM;
				
				System.out.println(prefixSum);
			}
		}
	}
}

File: ACM/tree/p2042. 구간 합 구하기.py
import sys
input = sys.stdin.readline

N, M , K = map(int, input().split())
MIN_NUM = -2 ** 63
MAX_NUM = 2 ** 63 - 1

k = 0
while N > 2 ** k:
    k += 1

tree_size = 2 ** (k+1)
tree = [0 for _ in range(tree_size)]

# Create Leaf Node
data_start_idx = tree_size // 2
for i in range(N):
    tree[data_start_idx + i] = int(input())

# Create Tree(Parent) Node
temp_start_idx = data_start_idx
temp_cnt = tree_size // 4
while temp_cnt > 0:
    parent_idx = temp_start_idx // 2
    for i in range(temp_cnt):
        tree[parent_idx + i] = tree[temp_start_idx + 2*i] + tree[temp_start_idx+1 + 2*i]
    temp_cnt = temp_cnt // 2
    temp_start_idx = temp_start_idx // 2

for i in range(M + K):
    a, b, c = map(int, input().split())

    start_index = data_start_idx + b - 1
    if a == 1:
        tree[start_index] = c
        parent_idx = start_index // 2
        while parent_idx >= 1:
            left_child_idx = parent_idx * 2
            right_child_idx = left_child_idx + 1
            tree[parent_idx] = tree[left_child_idx] + tree[right_child_idx]
            parent_idx = parent_idx // 2
    else:
        end_index = data_start_idx + c - 1
        pocket = []
        while end_index >= start_index:
            if start_index % 2:
                pocket.append(tree[start_index])
            if not end_index % 2:
                pocket.append(tree[end_index])
            start_index = (start_index + 1) // 2
            end_index = (end_index - 1) // 2

        prefix_sum = sum(pocket)
        if prefix_sum > MAX_NUM:
            prefix_sum = MAX_NUM
        elif prefix_sum < MIN_NUM:
            prefix_sum = MIN_NUM
        print(prefix_sum)